src/rgw/rgw_op.cc

```diff
@@ -1390,6 +1390,16 @@ int RGWOp::do_aws4_auth_completion()
 {
   ldpp_dout(this, 5) << "NOTICE: call to do_aws4_auth_completion"  << dendl;
   if (s->auth.completer) {
+#if 1
+    /* XXXX check props */
+    s->auth.completer->kv_list([this](string_view k, string_view v) {
+      ldpp_dout(this, 10) << __func__
+			  << " completer dump properties: "
+			  << "k: " << k << " v: " << v
+			  << dendl;
+      return 0;
+    });
```

Comment:
I don't really know whether pushing the headers up into s.info.env is really all that great idea, and so I don't see why you need to prescriptively decide that.  the potential advantage is that it reuses, yes, the memory in the completer, and why would it not want to?  my intuition about the big env is that it's a "blackboard"--a support for scribble-on coding style that I think I would prefer we get away from.  I could proxy radoslaw's argument for why not make that linkage too (aggressively prohibit interaction between completer and req_state);  I don't see the strong requirement to simplify that you are seeing, just yet

Reply:
the key-value interface is more general than it needs to be, but, it's efficient.

Reply:
(I find it sort of annoying to write, I originally had trailer_vec on the stack, and I moved it to elminate the string copies.  I don't see why the extra stack ops and small copies are preferable...)

Reply:
I don't agree that moving additional parsing logic into complete makes it easier for *me* to understand--to me, it's helpful to comprehend the recognition and one shot (recv_body)

Reply:
oh, and move on to checksums themselves, which are needed to provide, basically any check on what is being sent with STREAMING-UNSIGNED-PAYLOAD-TRAILER

Reply:
> I don't really know whether pushing the headers up into s.info.env is really all that great an idea, and so I don't see why you need to prescriptively decide that.

regarding the header representation, i was thinking ahead to the checksum comparison logic in `RGWPutObj::execute()` which would happen around where we finalize the etag checksum (https://github.com/ceph/ceph/blob/de5d8cae/src/rgw/rgw_op.cc#L4461-L4468). there we'll need to support both methods - where `x-amz-checksum-*` is either sent as a trailing header, or with the initial request headers

trailing headers also show up in http/2 and /3, and because they have the same semantics as normal request headers, i think rgw should expose them to RGWOps the same way

so i was starting from the assumption that we'd need to copy the trailing headers into RGWEnv anyway, and thinking about the Completer interface from that perspective. but if we do end up with special cases for trailing checksum headers in `RGWPutObj::execute()`, then i agree that the existing design provides efficient access via `completer->get("x-amz-checksum-crc32")`

Reply:
is there something to discuss about the current env implementation and available optimizations or restructuring available for it?

Reply:
> is there something to discuss about the current env implementation and available optimizations or restructuring available for it?

yeah, i've been using https://tracker.ceph.com/issues/57646 as a placeholder for that. the beast frontend parses request headers into [beast::http::fields](https://www.boost.org/doc/libs/1_83_0/libs/beast/doc/html/beast/ref/boost__beast__http__fields.html) then copies everything from there into `RGWEnv`. i would love to expose that beast container for everything - perhaps optimized with a custom per-request allocator. i think the hard part will just be tracking down all of the places that rely on the mangled header names, then figuring out backward-compat with the ones that get persisted to rados (like x-amz-meta- headers in xattrs)

Reply:
that and case cleanup would make me like this a lot more